java/org/contikios/cooja/serialsocket/SerialSocketServer.java

```diff
@@ -388,8 +387,29 @@ public boolean startServer(int port) {
           incomingDataHandler.start();
 
           /* Observe serial port for outgoing data */
-          serialDataObserver = new SerialDataObserver();
-          serialPort.addSerialDataObserver(serialDataObserver);
+          DataOutputStream outStream;
+          try {
+            outStream = new DataOutputStream(clientSocket.getOutputStream());
+          } catch (IOException ex) {
+            logger.error("Failed opening client socket output stream:", ex);
+            // FIXME: this should fail, not continue and produce invisible updates.
+            outStream = null;
+          }
+          final var out = outStream;
+          // FIXME: update code to use data directly.
+          serialPort.getSerialDataTriggers().addTrigger(this, serialDataObserver = (event, data) -> {
+            if (out == null) {
+              return;
+            }
+            try {
+              out.write(serialPort.getLastSerialData());
```

Comment:
Why not use `data` as it is provided by the trigger mechanism? Same in SerialSocketClient. This would allow `getLastSerialData()` to be deprecated or removed as it is more or less only useful from within a notification.

Reply:
That is the FIXME on line 399.

I want to remove the observers first, the API is easier to update when everything uses EventTriggers.